helper/impls/helper-impl-14/src/main/java/com/linkedin/avroutil1/compatibility/avro14/parsing/ValidatingGrammarGenerator.java

```diff
@@ -0,0 +1,153 @@
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * The class that generates validating grammar.
+ */
+public class ValidatingGrammarGenerator {
+  protected final static Set<Schema.Type> NAMED_TYPES = Collections.unmodifiableSet(new HashSet<>(
+          Arrays.asList(Schema.Type.ENUM, Schema.Type.FIXED, Schema.Type.RECORD)
+  ));
+
+  /**
+   * Returns the non-terminal that is the start symbol
+   * for the grammar for the given schema {@code sc}.
+   */
+  public Symbol generate(Schema schema) {
+    return Symbol.root(generate(schema, new HashMap<LitS, Symbol>()));
+  }
+
+  /**
+   * Returns the non-terminal that is the start symbol
+   * for the grammar for the given schema {@code sc}. If there is already an entry
+   * for the given schema in the given map {@code seen} then
+   * that entry is returned. Otherwise a new symbol is generated and
+   * an entry is inserted into the map.
+   * @param sc    The schema for which the start symbol is required
+   * @param seen  A map of schema to symbol mapping done so far.
+   * @return      The start symbol for the schema
+   */
+  public Symbol generate(Schema sc, Map<LitS, Symbol> seen) {
+    switch (sc.getType()) {
+    case NULL:
+      return Symbol.NULL;
+    case BOOLEAN:
+      return Symbol.BOOLEAN;
+    case INT:
+      return Symbol.INT;
+    case LONG:
+      return Symbol.LONG;
+    case FLOAT:
+      return Symbol.FLOAT;
+    case DOUBLE:
+      return Symbol.DOUBLE;
+    case STRING:
+      return Symbol.STRING;
+    case BYTES:
+      return Symbol.BYTES;
+    case FIXED:
+      return Symbol.seq(new Symbol.IntCheckAction(sc.getFixedSize()),
+          Symbol.FIXED);
+    case ENUM:
+      return Symbol.seq(new Symbol.IntCheckAction(sc.getEnumSymbols().size()),
+          Symbol.ENUM);
+    case ARRAY:
+      return Symbol.seq(Symbol.repeat(Symbol.ARRAY_END, generate(sc.getElementType(), seen)),
+          Symbol.ARRAY_START);
+    case MAP:
+      return Symbol.seq(Symbol.repeat(Symbol.MAP_END,
+              generate(sc.getValueType(), seen), Symbol.STRING),
+          Symbol.MAP_START);
+    case RECORD: {
+      LitS wsc = new LitS(sc);
+      Symbol rresult = seen.get(wsc);
+      if (rresult == null) {
+        Symbol[] production = new Symbol[sc.getFields().size()];
+
+        /**
+         * We construct a symbol without filling the array. Please see
+         * {@link Symbol#production} for the reason.
+         */
+        rresult = Symbol.seq(production);
+        seen.put(wsc, rresult);
+
+        int i = production.length;
+        for (Field f : sc.getFields()) {
+          production[--i] = generate(f.schema(), seen);
+        }
+      }
+      return rresult;
+    }
+    case UNION:
+      List<Schema> subs = sc.getTypes();
+      Symbol[] symbols = new Symbol[subs.size()];
+      String[] oldLabels = new String[subs.size()];
```

Comment:
actual change - 2 sets of labels (strings) instead of a single set - one for "simple" names (old style), another for FQCNs